teps/0060-remote-resource-resolution.md

```diff
@@ -239,13 +242,94 @@ users can leverage those shared pipelines.
+
+### Proposal 2: Disable Tekton Pipelines' "In-Tree" Resolution Logic With a Deployment Arg
+
+A problem we face with introducing an experimental resolver controller is that
+any in-cluster process watching for `TaskRun` and `PipelineRun` resources will
+race with Tekton Pipelines to process them as they're submitted to the kube
+API.
+
+In order to support an experimental controller independently performing
+resolution we could add an arg to Tekton Pipelines' controller (e.g.
+`--experimental-disable-in-tree-resolution`) that completely disables the
+built-in resolution logic.
+
+This is explicitly _not_ being proposed as a feature flag because 1) it is not
+intended to migrate through to production-readiness and 2) an arg change
+requires a hard restart of the controller pods, underlining the "experimental
+mode" that the user is electing to put their Tekton Pipelines controller in to.
+We can also print a warning when the controller boots up with this arg
+included.
+
+### Proposal 3: Add `status.resolvedMeta` to `TaskRuns` and `PipelineRuns`
+
+When the proposed experimental resolution controller sees an unresolved
+`TaskRun` or `PipelineRun` it will read the resource and then resolve whatever
+ref, bundle or inline document is included. The resolved `spec` will then be
+embedded in the `status.taskSpec` field for `TaskRuns` or the
+`status.pipelineSpec` field for `PipelineRuns`. Once those fields are populated
+Tekton Pipeline's reconcilers will be able to pick them up and begin working
+with the specs.
+
+We will take this approach because it's entirely backwards-compatible with our
+existing resolution and reconcilers: today, the reconcilers already copy
+resolved resources into the `status` block of `TaskRuns` and `PipelineRuns` so
+that changes in the underlying resources don't affect runs that are already
+in-flight. By observing the same protocol it means that the logic changes in
+our Tekton Pipelines' reconcilers is minimized and the same "audit trail"
+property is maintained.
+
+Unfortunatley there's a wrinkle: our existing reconcilers copy labels and
+annotations from the resolved `Task` or `Pipeline` to the `TaskRun` or
+`PipelineRun`. By splitting the resolution logic in to its own external
+machinery we are introducing a barrier between resolution and metadata
+duplication: the "actuation reconcilers" never see the fully-resolved resource,
+only the `spec` passed via the `status`, so they're unable to perform the
+metadata copy. The end result is that the "resolver reconcilers" are also
+responsible for performing this metadata copy.
+
+A lot of the conversation with this TEP so far has revolved around the fact
+that resolution might be something we want to expose not just for Pipelines but
+also other Tekton projects. Having the metadata copy performed by an external
+controller doesn't make a ton of sense in that case: metadata copying is a
+feature/requirement of Tekton Pipelines, not a general requirement for
+resolution.
```

Comment:
could you explain a bit more about this? im not totally following because:

1. afaik the experimental resolution controllers know enough about pipelineruns + taskruns to modify them directly, so in that sense they're not generic enough to use for anything else (maybe im missing something tho)
2. metadata like labels and annotations is pretty universal across k8s resources, so in that sense having some logic around this even in a very generic tool seems reasonable

Reply:
The resolution controller I'm proposing for the experimental repo would aim to deliver the functionality that we've been discussing through TEP-0060, which has now largely boiled down to: (1) It should be a separate project (2) It should process the resources before Pipelines sees them and (3) it shouldn't be Pipelines-specific.

It's true that metadata is a universal feature of objects in k8s. But the behaviour where metadata is copied verbatim from Tasks to TaskRuns and Pipelines to PipelineRuns is quite specifically Tekton Pipelines I think?  Let's say we open the system up in such a way that anyone can write a Resolver. Should we require their resolver to also perform this Pipelines-specific metadata copying action?  I guess I'm not convinced that this is an action we want left to external resolvers to perform. Those labels / annotations can trip up the pipelines controller when they're missing and are part of our backwards-compat story I think.

Reply:
> But the behaviour where metadata is copied verbatim from Tasks to TaskRuns and Pipelines to PipelineRuns is quite specifically Tekton Pipelines I think?

i wonder if one of the reasons we're copying around metadata like this is because we needed a better way to query pipelineruns and taskruns - which maybe is the role that tekton results fills now? might be something we want to consider for v1

>  I guess I'm not convinced that this is an action we want left to external resolvers to perform.

in that case im leaning more toward the approach where we embed the entire resolved yaml document (or as much of it as we know the controller needs) 🤔

Reply:
Yeah, possibly the biggest reason for the metadata duplication is about querying the cluster for PipelineRuns associated to Pipelines (and TaskRuns to Tasks). I definitely agree that Results might be a better host for this kind of relational data.

Just to throw an additional wrinkle (because I'm also generally in favor of embedding the entire yaml doc): if we go the route of embedding the entire resolved yaml I imagine there will be some amount of time where we need to store both `pipelineSpec` and (hypothetical) `pipelineYAML` in single PipelineRuns' `status` for compatibility's sake. In turn I wonder if this would push PipelineRuns much more quickly towards etcd's size limits when very large Pipelines are involved. wdyt?

Reply:
Sorry, didn't quite finish my though inre: metadata duplication.  While querying associations is one very big use case I think there are also probably others and it would probably behoove us to enumerate those and make sure they're accounted for before we decide to drop the copying behaviour.

Reply:
> I imagine there will be some amount of time where we need to store both pipelineSpec and (hypothetical) pipelineYAML in single PipelineRuns' status for compatibility's sake.

hmm I'm trying to think through the scenario where we'd need to do this - all I can think of is if the controller is downgraded during pipelinerun execution?

The current controller expects to find a pipeline spec in the pipelinerun status - the updated version would expect to find the entire pipelineYAML but could also support the pipeline spec if it found it (we could wait until we're outside the compatibility window to remove that support) - but most cases wouldnt require actually storing both i dont think?

scenarios i can think of:

1. upgrade from "pipeline spec" controller to new version:
    * completed pipelineruns would contain "pipeline spec" in status, doesnt matter cuz they're done
    * in progress pipelineruns would be using "pipeline spec" which would be okay as long as the new version still supported it
    * new pipelineruns would start using "pipelineYAML"
2. downgrade from new version of the controller to "pipeline spec" controller
    * completed pipeline runs contain "pipelineYAML", doesnt matter cuz they're done (possible that on controller start up it gets upset about unexpected fields?)
    * in progress pipeline runs would be in trouble - the old controller wouldn't know how to read "pipelineYAML" out of their specs
    * new pipeline runs would use "pipeline spec"

Do we need to support the downgrade case in (2)? (specifically: support runs which execute during the update?)

Reply:
I was specifically thinking of users who have written code that expects the `pipelineSpec` to be embedded in the status rather than scenarios where the system itself might be caught in an intermediate state. But it sounds like you don't think this is a situation we need to worry about?

Reply:
OK, I've updated the proposal to pass the entire resolved yaml via a `taskYaml` field for taskruns and a `pipelineYaml` field for pipelineruns.

FYI @vdemeester and @dibyom - this is a change to the proposal so probably worth giving it one more look to confirm you're happy to keep approving.

Reply:
Resolving both `metadata` and `spec` under a `taskYAML` or `pipelineYAML` field sounds good to me.

Reply:
> users who have written code that expects the pipelineSpec to be embedded in the status

ah that's a good point, i was only thinking about this from the controller's perspective - i always forget about status as part of the api 😅 

you're probably right, it's probably best to support both - maybe we can use (even more) config options to control whether we do that or not?